Extend `RUIProvider` with `globalProps` to pass props to all UI and layout components using provider, create Component Customization page in documentation (#215)
react-ui-org/react-ui: src/docs/customize/component-customization.mdx

---
name: Component Customization
menu: Customize
route: /customize/component-customization
---

# Component Customization

> ## Global props

Reviewer: Please stay consistent in capitalizing: _Global Props_, also see other headings below.

import { Playground } from 'docz'
import {
  Toolbar,
  ToolbarItem,
} from '../../lib/components/layout/Toolbar';
import Button from '../../lib/components/ui/Button';
import SelectField from '../../lib/components/ui/SelectField';
import TextField from '../../lib/components/ui/TextField';
import { RUIProvider } from '../../lib/provider';

> React UI often provides more variants of components, so a user can

Reviewer: > Many React UI components are available in more variants so users can pick what suits them the best.

…?

pick the most suitable variant.

> There two ways to configure component:

Reviewer: > There are two ways …

1. Pass props directly to a specific component.
2. Wrap application (or its part) with `RUIProvider` and pass arbitrary props for

> a specific component through `globalProps` prop.

Reviewer: I like Martin's wording of the second option but I wouldn't delete the first one. Only number 2 is in scope of this issue but number 1 is generally still valid, IMO.

> a specific component through `globalProps` prop.

Reviewer 2: How about adding the information to the end:
```
…
and passing in the common props through `globalProps` prop.
With or without using `RUIProvider` the component can be configured by passing in the props directly.
The props passed in directly always take precedence over props coming from `RUIProvider`.
```

I find the overloading bit important.

### Basic Usage

To define global props, you need to import `RUIProvider` first:

```js
import { RUIProvider } from '@react-ui-org/react-ui';
```

> Then wrap application (or its part) with `RUIProvider` with `globalProps` prop object,
> where its keys are names of components and its values are objects with arbitrary
> props you want to pass to a specific component.
> 
> Keys conform to actual names of components:

Reviewer 2: ```suggestion
Then wrap application or its part with `RUIProvider` with defined `globalProps` attribute holding an object.
The keys are names of the components and its values are objects with arbitrary
props you want to pass to the specified components.

```

Reviewer: and their values

```jsx
<RUIProvider globalProps={{
  Button: { priority: 'filled' },
  TextField: { variant: 'filled' },
  SelectField: { variant: 'filled' },
}}>
  //...
</RUIProvider>
```

> See working example:

Reviewer 2: I like the working example :heart:

Reviewer: Yep this is cool!

    const [variant, setVariant] = React.useState('filled');
    return (
      <RUIProvider
        globalProps={{
          Button: { priority: variant },
          TextField: { variant },
          SelectField: { variant },
        }}
      >
        <Toolbar align="bottom">
          <ToolbarItem>
            <SelectField
              changeHandler={(e) => setVariant(e.target.value)}
              id="variant"
              label="Select variant of Select Field"
              options={[
                {
                  label: 'filled',
                  value: 'filled',
                },
                {
                  label: 'outline',
                  value: 'outline',
                },
              ]}
              value={variant}
            />
          </ToolbarItem>
          <ToolbarItem>
            <TextField
              id="my-text-field"
              label="Text Field"
            />
          </ToolbarItem>
          <ToolbarItem>
            <Button
              clickHandler={() => {}}
              id="my-button"
              label="Button"
            />
          </ToolbarItem>
        </Toolbar>
      </RUIProvider>
    );
  }}
</Playground>

> ### Best practices

Reviewer 2: I'm not sure about this section.

For example, when my entire project is just one `<Form>` I can imagine something  like:
```
<RUIProvider globalProps={{
  TextField: { clickHandler: setValue },
}}>
```
I would prefer to leave this up to the discretion of the developer using this feature.

Reviewer: Maybe a different wording would help? Something like _however there are component props that are usually unique to their instance_…

Reviewer 2: But why?

All props are handled exactly the same way. There is no technical reason for it. As far as non technical reasons that depends on the application. Sometimes it might be a terrible idea and sometimes it can be a simple and elegant solution.  Only the application author knows what props are to be unique and what props  can be shared in the given context.

`usually unique` always depends for whom... if all I ever do are single page three field forms over and over then my definition is very different from someone doing complex administrations.

Global props is a powerful tool that allows you to pass arbitrary props
to every component you want. It has been made to simplify repeatedly
set props that are common for your whole project.

For that reason **we do recommend to set only global props which have
impact on appearance**. Do not use it for props that are used to be
unique, e.g `clickHandler`,`closeHandler`, `disabled`, `id`, `label`,  etc.

## Translations

> Some components may contain tooltips which improve components' accessibility.

Reviewer 2: ```suggestion
Some components may contain texts which improve components' accessibility.
```

Reviewer: Maybe _visually hidden texts_ then…?

Reviewer 2: They are hidden now, but they could be visible in the future.

I think that in this case being more specific does not help in any way. All texts, hidden or visible, have to be translated and when setting up react-ui in multi-language setting I do need to know this information.

All texts are in English by default and can be translated to other languages.

> ### Translation Approaches

Reviewer 2: I would consider moving translation documentation to a separate page. Its close relatation to globalProps is in my view only an implementation detail.

Reviewer: Yes, @bedrich-schindler please create a separate page for this and link the pages through.
